pkg/capture/metrics.go

```diff
@@ -12,108 +12,217 @@ const (
+		promRotationDuration: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Namespace: config.ServiceName,
+			Subsystem: captureManagerSubsystem,
+			Name:      "rotation_duration_seconds",
+			Help:      "Total flow map rotation time, aggregated across all interfaces",
+			// rotation is significantly faster than the writeout. Hence the small buckets
+			Buckets: []float64{0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 0.75, 1},
+		}),
+	}
+
 	prometheus.MustRegister(
-		promPacketsProcessed,
-		promPacketsDropped,
-		promBytes,
-		promPackets,
-		promGlobalBufferUsage,
-		promNumFlows,
-		promCaptureIssues,
-		promInterfacesCapturing,
-		promRotationDuration,
+		metrics.promPacketsProcessed,
+		metrics.promPacketsDropped,
+		metrics.promBytes,
+		metrics.promPackets,
+		metrics.promGlobalBufferUsage,
+		metrics.promNumFlows,
+		metrics.promCaptureIssues,
+		metrics.promInterfacesCapturing,
+		metrics.promRotationDuration,
 	)
+
+	return metrics
+}
+
+func (m *Metrics) ObservePacketsProcessed(iface string) prometheus.Counter {
+	if m.trackIfaces {
+		return m.promPacketsProcessed.WithLabelValues(iface)
+	}
+	return m.promPacketsProcessed.WithLabelValues()
+}
+
+func (m *Metrics) ObservePacketsDropped(iface string) prometheus.Counter {
+	if m.trackIfaces {
+		return m.promPacketsDropped.WithLabelValues(iface)
+	}
+	return m.promPacketsDropped.WithLabelValues()
+}
+
+func (m *Metrics) ObserveBytesTotal(iface, direction string) prometheus.Counter {
+	if m.trackIfaces {
+		return m.promBytes.WithLabelValues(iface, direction)
+	}
+	return m.promBytes.WithLabelValues(direction)
+}
+
+func (m *Metrics) ObservePacketsTotal(iface, direction string) prometheus.Counter {
+	if m.trackIfaces {
+		return m.promPackets.WithLabelValues(iface, direction)
+	}
+	return m.promPackets.WithLabelValues(direction)
+}
+
+func (m *Metrics) ObserveGlobalBufferUsage(iface string) prometheus.Gauge {
+	if m.trackIfaces {
+		return m.promGlobalBufferUsage.WithLabelValues(iface)
+	}
+	return m.promGlobalBufferUsage.WithLabelValues()
+}
+
+func (m *Metrics) ObserveNumFlows(iface string) prometheus.Gauge {
+	if m.trackIfaces {
+		return m.promNumFlows.WithLabelValues(iface)
+	}
+	return m.promNumFlows.WithLabelValues()
+}
+
+func (m *Metrics) ObserveCaptureIssues(iface, issue_type string) prometheus.Counter {
+	if m.trackIfaces {
+		return m.promCaptureIssues.WithLabelValues(iface, issue_type)
+	}
+	return m.promCaptureIssues.WithLabelValues(issue_type)
+}
+
+func (m *Metrics) ObserveNumIfacesCapturing() prometheus.Gauge {
+	return m.promInterfacesCapturing
+}
+
+func (m *Metrics) ObserveRotationDuration() prometheus.Histogram {
+	return m.promRotationDuration
 }
 
 // ResetCountersTestingOnly allows to externally reset all Prometheus counters (e.g. for
 // testing purposes or in order to manually reset all of them)
 // This method must not (and cannot) be called outside of testing
-func ResetCountersTestingOnly() {
+func (m *Metrics) ResetCountersTestingOnly() {
```

Comment:
Cherry on top: Avoids the even uglier resetting of global counters from tests and now makes it a member of the type (still keeping it `testing` only as there currently is no other use case and it can have mean side effects.

Reply:
+1 for the public comment advising caution